File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>

using namespace std;

struct ROR{
    int x, y, cnt;
    // ROR(int x, int y, int cnt) : x(x), y(y), cnt(cnt) {}
};

int answer;
int N, M;
bool visited[101][101];

int dx[4] = {0, 0, 1, -1};
int dy[4] = {1, -1, 0, 0};

int solution(vector<vector<int>> maps)
{
    answer = -1;
    N = maps.size();
    M = maps[0].size();
    
    queue<ROR> q;
    q.push({0, 0, 1});
    visited[0][0] = true;
    
    while(!q.empty()){
        ROR tmp = q.front();
        q.pop();
        
        if(tmp.x == N - 1 && tmp.y == M - 1){
            answer = tmp.cnt;
            break;
        }
        
        for(int i = 0; i < 4; ++i){
            int nx = tmp.x + dx[i];
            int ny = tmp.y + dy[i];
            
            if(nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
            if(maps[nx][ny] == 0 || visited[nx][ny]) continue;
        
            visited[nx][ny] = true;
            q.push({nx, ny, tmp.cnt + 1});
        }
    }
    
    return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    int[][] dxdy = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    
    public int solution(int[][] maps) {
        int answer = -1;
        
        int n = maps.length;
        int m = maps[0].length;
        
        int[][] dist = new int[n][m];
        
        Queue<int[]> q = new ArrayDeque<>();
        
        q.offer(new int[] {0, 0});
        dist[0][0] = 1;
        
        while(!q.isEmpty()) {
            int[] cur = q.poll();

            if(cur[0] == n - 1 && cur[1] == m - 1) {
                answer = dist[cur[0]][cur[1]];
                break;
            }
            
            for(int i = 0; i < 4; i++) {
                int nx = cur[0] + dxdy[i][0];
                int ny = cur[1] + dxdy[i][1];
                
                if(nx < 0 || nx >= n || ny < 0 || ny >= m) {
                    continue;
                }
                
                if(maps[nx][ny] == 0) {
                    continue;
                }
                
                if(dist[nx][ny] != 0) {
                    continue;
                }
                
                dist[nx][ny] = dist[cur[0]][cur[1]] + 1;
                q.offer(new int[] {nx ,ny});
            }
        }
        
        return answer;
    }
}